Add Soviet 09

Smittytron opened the pull request:
A couple notes for this one:

- The north allied base does not have enough power starting power due to different power values from the original. I've added a small power plant in the top right after discussion on discord.

- I have the blue allied team (greece) not producing units as they would likely just blow up on their own mines while attacking. I have water reinforcements attacking instead.

The original mission was very easy to speed run if you know way the truck goes, and it's the same way in this PR. I could try to add AA guns and units where they don't belong to prevent this, but at the moment I'm keeping it close to the original.


~~Edit: Depends on #18122~~

Smittytron committed:
Add Soviet 09

Smittytron commented:
Also depends on #18308

Smittytron commented:
Rebased and updated to work with #18308. No longer depends on #18122.

abcdefg30 commented on mods/ra/maps/soviet-09/soviet09-AI.lua:
```diff
@@ -0,0 +1,190 @@
+end
+
+ProduceInfantry = function()
+	if (GermanyTent.IsDead or GermanyTent.Owner ~= Germany) then
```
The parenthesis can be removed.

abcdefg30 commented on mods/ra/maps/soviet-09/soviet09.lua:
```diff
@@ -0,0 +1,103 @@
+	end)
+	
+	DestroyTruck = USSR.AddPrimaryObjective("Destroy the stolen convoy truck.\nDo not let it escape.")
+	BeatUSSR = Greece.AddPrimaryObjective("Defeat the Soviet forces.")
```
Needs to use the new Api.

Smittytron commented:
Updated

Smittytron commented:
Added an objective to defend the forward command center.

Smittytron commented:
Rebased

abcdefg30 commented on mods/ra/maps/soviet-09/soviet09-AI.lua:
```diff
@@ -0,0 +1,190 @@
+
+	Utils.Do(AttackGroup, function(unit)
+		if not unit.IsDead then
+			Trigger.OnIdle(unit, unit.Hunt)
```
This should likely use `IdleHunt` too? `Utils.Do(AttackGroup, IdleHunt)`

abcdefg30 commented on mods/ra/maps/soviet-09/soviet09-AI.lua:
```diff
@@ -0,0 +1,190 @@
+WTransUnits = { { "2tnk", "1tnk", "1tnk", "e3", "e3" }, { "2tnk", "2tnk", "2tnk"  } }
+WTransDelays =
+{
+	easy = 5,
```
I suppose we want to do the same here as in `08` and define the minutes directly.

abcdefg30 commented on mods/ra/maps/soviet-09/soviet09.lua:
```diff
@@ -0,0 +1,108 @@
+	Greece.Cash = 50000
+	Germany.Cash = 50000
+
+	if USSR.HasNoRequiredUnits() then
```
This shouldn't be needed since you already lose when you lose the command center, don't you? We can then remove the `BeatUSSR` objective too.

Smittytron commented:
Updated

cjshmyr reviewed (changes requested):
Played through on Normal, was fun!

The convoy truck at the end of the mission never moved for me. After I destroyed it, the objective was marked as completed, but the other primary objective (protect the command center) was still incomplete, so I couldn't finish the mission.

Potential polish:
- I would space out the chinook drops a bit more (change from 30s-60s to 60s-90s), they came fast and often, which was a bit overwhelming. But this is just my opinion.
- The originals had a camera reveal the base when you entered it by foot (although for a brief second, so you could find the convoy truck before it tried to escape). This might be good polish to add if the truck escapes before we can see it, or get too close to it.

cjshmyr commented:
These are some general comments related to missions in general, not necessarily this PR.
- Do we plan on having mobile flak buildable at some point? They are enemies in early allies missions.
- In the originals the command center is repairable, but in here (and Soviet 02) it is not. Personally it felt odd I could not repair a base building, but I won't argue either way.

Smittytron commented:
Updated. Truck fleeing trigger fixed and protect command center objective will mark completed on truck kill. I set longer timers on easy and normal for the Chinook drops. I also increased the build delay for replacing Apaches. Added a 5 second camera by the truck when you make it flee.


> Do we plan on having mobile flak buildable at some point? They are enemies in early allies missions.

Not sure, I don't plan on it personally. Flaks would make dealing with AI air attacks incredibly easy.

Smittytron committed:
Allow repairable tech buildings in campaign-rules

Smittytron commented:
> In the originals the command center is repairable, but in here (and Soviet 02) it is not. Personally it felt odd I could not repair a base building, but I won't argue either way.

Added a commit to allow repair of all tech buildings in campaign-rules, since we'll run into the same issue with the others.

cjshmyr commented on mods/ra/maps/soviet-09/soviet09.lua:
```diff
@@ -0,0 +1,109 @@
+end
+
+Trigger.OnEnteredProximityTrigger(TruckAlarm.CenterPosition, WDist.FromCells(13), function(actor, triggerflee)
+	if actor.Owner == USSR and not (actor.Type == "badr" or actor.Type == "u2") then
```
Doesn't fully exclude spy planes or badgers, this will:

`	if actor.Owner == USSR and not (actor.Type == "badr" or actor.Type == "u2" or actor.Type == "camera.paradrop" or actor.Type == "camera.spyplane") then`

I don't know of a better way to exclude cameras from proxy triggers other than a new trait, or a re-usable proxy exclusion check in campaign-global.

abcdefg30 replied:
Fwiw, it would be better to use `~=` over `not ==`:
`if actor.Owner == USSR and actor.Type ~= "badr" and actor.Type ~= "u2" and actor.Type ~= "camera.paradrop" and actor.Type ~= "camera.spyplane" then`

I'm also not sure if we shouldn't have the truck start escaping when parabombs or paratroopers come in.

Smittytron replied:
I don't have parabombs available in this one. Paratroopers should trigger the truck when they jump out. I want to avoid a situation where the player drops paras on their side of the map and the plane passes over the trigger on the way out.

abcdefg30 replied:
> I don't have parabombs available in this one.

Interesting. What was the idea here, that it is too strong against the truck?

You're right about the planes passing, the better solution if we want it would be to drop the checks for the camera I guess.

Smittytron replied:
> What was the idea here, that it is too strong against the truck?

Yeah, then again I think everything is. Really no way to make this mission speed-run cheese proof.

cjshmyr commented on mods/ra/maps/soviet-09/rules.yaml:
```diff
@@ -0,0 +1,121 @@
+	Buildable:
+		Prerequisites: ~disabled
+
+HPAD:
```
Stole an Allied conyard and wasn't able to build a helipad, though I did capture one next to it. Was it disabled intentionally? (tech level?)

Smittytron replied:
I had intended to address this in my heli pr with an allies hpad and soviet hpad. I guess I could make the prereq allied tech.

Smittytron commented:
Updated, changed hpad prereqs to `~structures.allies` and added spyplane cam to not trigger the truck fleeing function. I'll leave the paradrop cam since we want the truck to flee if you drop right on top of the truck anyways.

> Added a commit to allow repair of all tech buildings in campaign-rules

On second thought, civ buildings default to tech buildings for whatever reason, so this creates a travis nightmare. For now I've just put the override on `FCOM`

cjshmyr commented on mods/ra/maps/soviet-09/soviet09.lua:
```diff
@@ -0,0 +1,109 @@
+	end)
+end
+
+Trigger.OnEnteredProximityTrigger(TruckAlarm.CenterPosition, WDist.FromCells(13), function(actor, triggerflee)
```
I would shrink this a bit? In the originals the trigger fired when you were near the sandbags, but we're further away. Ending clip for reference: https://www.twitch.tv/hambtv/clip/DifferentBovineParrotCopyThis

Smittytron replied:
Updated, I moved the waypoint up by one and reduced the distance by two. I can't match it up perfectly without getting in too close from the sides. I want a proximity trigger instead of footprint here to account for paratroopers.

cjshmyr replied:
Makes sense, sounds good

cjshmyr reviewed (commented):
Finished again on hard, awesome job.

obrakmann commented on mods/ra/rules/campaign-rules.yaml:
```diff
@@ -67,6 +67,16 @@ FCOM:
+		Sequence: repair
+		Position: Center
+		Palette: player
+		IsPlayerPalette: True
```
Does this really belong in campaign-rules.yaml?

Smittytron replied:
Added this in response to a comment from @cjshmyr:

> In the originals the command center is repairable, but in here (and Soviet 02) it is not. Personally it felt odd I could not repair a base building, but I won't argue either way.

If we add this here at all we'll want it repairable in all missions.

obrakmann commented on mods/ra/maps/soviet-09/rules.yaml:
```diff
@@ -0,0 +1,121 @@
+Player:
+	PlayerResources:
+		DefaultCash: 8000
+
```
[On IRC, it was mentioned that the ending FMV cuts in too soon](https://logs.openra.net/?year=2020&month=07&day=05#23:41:30).  If you want you can increase the delay with the GameOverDelay setting on MissionObjectives like so:

```yaml:
Player:
	MissionObjectives:
		GameOverDelay: 3500
```

pchote replied:
I expect that this will impact other missions too?

IMO would be best to address this in a separate PR, maybe by adding this to the campaign-rules or even changing the default on the trait?

Smittytron replied:
> IMO would be best to address this in a separate PR, maybe by adding this to the campaign-rules or even changing the default on the trait?

Agreed, I've noticed this being abrupt in most if not all missions.

obrakmann reviewed (commented):
This is a nicely scripted mission, it was fun to play, :+1: 

Let us know whether you want to address these two comments, otherwise it's good to go, I think.

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/6b53d7169d3b27edf688fa397bff7f920da72114)
